### PR TITLE
MGMT-10816: Returning the actual resulting cluster platform

### PR DESCRIPTION
**Pull request by eliorerz**
- Make `none platform` as a valid platform
- When setting `UserManagedNetworking` (on cluster creation and on cluster update), if `UserManagedNetworking=true && Platform.Type == "baremetal"` than `cluster.Platform.Type` is set to `"none"`
- Allow creating and updating cluster with `Platform.Type = none`  
- `GetSupportedProvidersByHosts` returns also `none` platform alongside `baremetal`

## List all the issues related to this PR

- [x] New Feature 
- [ ] Enhancement 
- [x] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [x] Cloud
- [ ] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [x] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [x] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Assignees

/cc @eranco74 
/cc @avishayt 
/cc @gamli75

### CODE REVIEW

File: internal/bminventory/inventory.go [hunk 1]
```diff
@@ -1653,6 +1660,23 @@ func (b *bareMetalInventory) UpdateClusterNonInteractive(ctx context.Context, pa
 	return b.v2UpdateClusterInternal(ctx, params, NonInteractive)
 }
 
+func updateClusterNonePlatformParams(params installer.V2UpdateClusterParams, cluster *common.Cluster) {
+	if params.ClusterUpdateParams.UserManagedNetworking == nil || cluster.Platform == nil || cluster.UserManagedNetworking == nil {
```
Comment (eranco74) [inline internal/bminventory/inventory.go:1664]: This should be && no?

File: internal/bminventory/inventory.go [hunk 2]
```diff
@@ -1653,6 +1660,23 @@ func (b *bareMetalInventory) UpdateClusterNonInteractive(ctx context.Context, pa
...
+		return
+	}
+
+	if *params.ClusterUpdateParams.UserManagedNetworking && !*cluster.UserManagedNetworking {
+		if *cluster.Platform.Type == models.PlatformTypeBaremetal {
+			params.ClusterUpdateParams.Platform = &models.Platform{Type: common.PlatformTypePtr(models.PlatformTypeNone)}
+
+		} else if params.ClusterUpdateParams.Platform != nil && models.PlatformTypeBaremetal == *params.ClusterUpdateParams.Platform.Type {
+			params.ClusterUpdateParams.Platform = &models.Platform{Type: common.PlatformTypePtr(models.PlatformTypeNone)}
```
Comment (eranco74) [inline internal/bminventory/inventory.go:1673]: You are basically overriding a param provided by the users because it's conflicting with another param,
I think this deserves returning bad request, @avishayt?

File: internal/bminventory/inventory.go [hunk 3]
```diff
@@ -1653,6 +1660,23 @@ func (b *bareMetalInventory) UpdateClusterNonInteractive(ctx context.Context, pa
...
+		}
+	} else if params.ClusterUpdateParams.Platform != nil && *params.ClusterUpdateParams.Platform.Type == models.PlatformTypeNone {
+		params.ClusterUpdateParams.UserManagedNetworking = swag.Bool(true)
+	}
```
Comment (eranco74) [inline internal/bminventory/inventory.go:1677]: What about the case where user sends `UserManagedNetworking = false`, and `cluster.UserManagedNetworking = true`?
it seems that that case is ignored.
Also, if we are not to returning badRequest we should make it clear what param takes presedence Platform or UMN flag.

File: internal/bminventory/inventory.go [hunk 4]
```diff
@@ -1689,6 +1713,8 @@ func (b *bareMetalInventory) v2UpdateClusterInternal(ctx context.Context, params
 		return nil, common.NewApiError(http.StatusNotFound, err)
 	}
 
+	updateClusterNonePlatformParams(params, cluster)
```
Comment (eranco74) [inline internal/bminventory/inventory.go:1716]: Consider renaming e.g. `alignClusterPlatformParams`

File: internal/bminventory/inventory.go (hunk 2 above)
Comment (avishayt) [reply internal/bminventory/inventory.go:1673]: I agree.  If they selected user-managed networking and baremetal that's 400.

File: internal/bminventory/inventory.go (hunk 1 above)
Comment (eliorerz) [reply internal/bminventory/inventory.go:1664]: No, If one of those properties is not set it will not be able to update the platform/UMN

File: internal/bminventory/inventory.go (hunk 3 above)
Comment (eliorerz) [reply internal/bminventory/inventory.go:1677]: > What about the case where user sends UserManagedNetworking = false, and cluster.UserManagedNetworking = true?

You right, but I think we can handle this case instead `*params.ClusterUpdateParams.UserManagedNetworking == false && *cluster.Platform.Type == models.PlatformTypeNone`
WDYT?

Comment (eranco74) [reply internal/bminventory/inventory.go:1677]: Unsure how we can handle it though.

File: internal/bminventory/inventory.go [hunk 5]
```diff
@@ -1653,6 +1660,23 @@ func (b *bareMetalInventory) UpdateClusterNonInteractive(ctx context.Context, pa
 	return b.v2UpdateClusterInternal(ctx, params, NonInteractive)
 }
 
+func updateClusterNonePlatformParams(params installer.V2UpdateClusterParams, cluster *common.Cluster) {
+	if params.ClusterUpdateParams.UserManagedNetworking == nil || cluster.Platform == nil || cluster.UserManagedNetworking == nil {
+		return
+	}
+
+	if *params.ClusterUpdateParams.UserManagedNetworking && !*cluster.UserManagedNetworking {
```
Comment (eranco74) [inline internal/bminventory/inventory.go:1668]: if cluster.UserManagedNetworking is nil you can hit nil pointer dereference

File: internal/bminventory/inventory.go (hunk 1 above)
Comment (eranco74) [reply internal/bminventory/inventory.go:1664]: Hmmm, unsure I understand the logic here.
Don't we have a default for `cluster.Platform`? under what circumstances it will be nil?
Same question for `cluster.UserManagedNetworking`.

If `params.ClusterUpdateParams.UserManagedNetworking` is nil but `params.ClusterUpdateParams.Platform.Type == models.PlatformTypeNone` you would want to update `params.ClusterUpdateParams.UserManagedNetworking` to `swag.Bool(true)` no?

File: internal/bminventory/inventory.go (hunk 3 above)
Comment (eliorerz) [reply internal/bminventory/inventory.go:1677]: I think that in that case we should return to baremetal. The "duplication" of UMN and platform=none casing some edge cases that not that trivial to handle (e.g. selected vsphere platfrom with UMN=true and change back to baremetal, is it disabling the UMN?)

Comment (eranco74) [reply internal/bminventory/inventory.go:1690]: These 2 parameters are optional, if the user chooses to provide both with conflicting values we shouldn't overthink it and simply return bad request.
In case one was provided we should align the other param as well (within the boundary of None / Baremetal platform and not overriding any other platforms provided by the user).

File: internal/bminventory/inventory.go (hunk 2 above)
Comment (eliorerz) [reply internal/bminventory/inventory.go]: i updated it and now returning bed-request when `UserManagedNetworking=false and Platform=none or UserManagedNetworking=true and Platform=baremetal`

**Comment by eliorerz** [comment]
> Hmmm, unsure I understand the logic here.
Don't we have a default for cluster.Platform? under what circumstances it will be nil?
Same question for cluster.UserManagedNetworking.

@eranco74 You are right, I removed all the unnecessary `nil` "validations"

File: internal/bminventory/inventory.go (hunk 1 above)
Comment (eliorerz) [reply internal/bminventory/inventory.go]: @eranco74 You are right, I removed all the unnecessary nil "validations"

File: internal/bminventory/inventory.go [hunk 6]
```diff
@@ -458,6 +458,13 @@ func (b *bareMetalInventory) RegisterClusterInternal(
 		monitoredOperators = append(monitoredOperators, newOLMOperators...)
 	}
 
+	platform := params.NewClusterParams.Platform
+	if models.PlatformTypeBaremetal == *platform.Type && *params.NewClusterParams.UserManagedNetworking {
```
Comment (eranco74) [inline internal/bminventory/inventory.go:462]: params.NewClusterParams.UserManagedNetworking might be nil!

File: internal/bminventory/inventory.go [hunk 7]
```diff
@@ -458,6 +458,13 @@ func (b *bareMetalInventory) RegisterClusterInternal(
...
+		*platform.Type = models.PlatformTypeNone
```
Comment (eranco74) [inline internal/bminventory/inventory.go:463]: Didn't we agree that this case should return bad request?

File: internal/bminventory/inventory.go [hunk 8]
```diff
@@ -458,6 +458,13 @@ func (b *bareMetalInventory) RegisterClusterInternal(
...
+	} else if models.PlatformTypeNone == *platform.Type {
+		params.NewClusterParams.UserManagedNetworking = swag.Bool(true)
```
Comment (eranco74) [inline internal/bminventory/inventory.go:465]: I think you want to check that that params.NewClusterParams.UserManagedNetworking is nil before overriding it

File: internal/bminventory/inventory.go [hunk 9]
```diff
@@ -458,6 +458,13 @@ func (b *bareMetalInventory) RegisterClusterInternal(
 		monitoredOperators = append(monitoredOperators, newOLMOperators...)
 	}
 
+	platform := params.NewClusterParams.Platform
+	if models.PlatformTypeBaremetal == *platform.Type && *params.NewClusterParams.UserManagedNetworking {
+		*platform.Type = models.PlatformTypeNone
+	} else if models.PlatformTypeNone == *platform.Type {
```
Comment (eranco74) [inline internal/bminventory/inventory.go:464]: Consider moving all this to a function than handles the platform and UMN params.
e.g. `alignClusterPlatformParams`

File: internal/bminventory/inventory.go (hunk 6 above)
Comment (eranco74) [inline internal/bminventory/inventory.go:462]: Note that in line 421 `setDefaultRegisterClusterParams` will set the platform to Baremetal if the user didn't specify it.
Meaning whatever checks you are doing for the params input it should happen before we set defaults.

File: internal/bminventory/inventory.go [hunk 10]
```diff
@@ -1660,6 +1667,28 @@ func (b *bareMetalInventory) UpdateClusterNonInteractive(ctx context.Context, pa
...
+		if !*paramsUserManagedNetworking && platformType == models.PlatformTypeNone || *paramsUserManagedNetworking && models.PlatformTypeBaremetal == platformType {
+			return common.NewApiError(http.StatusBadRequest, errors.Errorf("Can't set %s platform with user-managed-networking %v", platformType, *paramsUserManagedNetworking))
+		}
+	}
+
+	if swag.BoolValue(paramsUserManagedNetworking) && *cluster.Platform.Type == models.PlatformTypeBaremetal {
+		params.ClusterUpdateParams.Platform = &models.Platform{Type: common.PlatformTypePtr(models.PlatformTypeNone)}
+	} else if params.ClusterUpdateParams.Platform != nil && *params.ClusterUpdateParams.Platform.Type == models.PlatformTypeNone {
+		params.ClusterUpdateParams.UserManagedNetworking = swag.Bool(true)
+	} else if paramsUserManagedNetworking != nil && !*paramsUserManagedNetworking && *cluster.Platform.Type == models.PlatformTypeNone {
```
Comment (eranco74) [inline internal/bminventory/inventory.go:1686]: This case (UMN=false, platform=None) should return bad request

File: internal/bminventory/inventory.go [hunk 11]
```diff
@@ -1660,6 +1667,28 @@ func (b *bareMetalInventory) UpdateClusterNonInteractive(ctx context.Context, pa
...
+	if paramsUserManagedNetworking != nil && params.ClusterUpdateParams.Platform != nil {
+		platformType := *params.ClusterUpdateParams.Platform.Type
+		if !*paramsUserManagedNetworking && platformType == models.PlatformTypeNone || *paramsUserManagedNetworking && models.PlatformTypeBaremetal == platformType {
+			return common.NewApiError(http.StatusBadRequest, errors.Errorf("Can't set %s platform with user-managed-networking %v", platformType, *paramsUserManagedNetworking))
+		}
+	}
+
+	if swag.BoolValue(paramsUserManagedNetworking) && *cluster.Platform.Type == models.PlatformTypeBaremetal {
+		params.ClusterUpdateParams.Platform = &models.Platform{Type: common.PlatformTypePtr(models.PlatformTypeNone)}
+	} else if params.ClusterUpdateParams.Platform != nil && *params.ClusterUpdateParams.Platform.Type == models.PlatformTypeNone {
```
Comment (eranco74) [inline internal/bminventory/inventory.go:1684]: So if platform == none it will just override params.ClusterUpdateParams.UserManagedNetworking to true?
why?

Comment (eliorerz) [reply internal/bminventory/inventory.go:1684]: It will set the platform_type to none and will update UserManagedNetworking=true.
The update will be later when UserManagedNetworking and platform is updated

File: internal/bminventory/inventory.go (hunk 10 above)
Comment (eliorerz) [reply internal/bminventory/inventory.go:1686]: Didn't we said we want to keep the current behavior?
If the user set the UMN=true that resulting platform=none, now updating UMN=false will return bad request.
In this case we will need to also change the UI behavior?

File: internal/bminventory/inventory.go [hunk 12]
```diff
@@ -1668,6 +1676,107 @@ func (b *bareMetalInventory) UpdateClusterNonInteractive(ctx context.Context, pa
...
+		if swag.BoolValue(userManagedNetworking) || (platform != nil && *platform.Type == models.PlatformTypeNone) {
+			return &models.Platform{Type: common.PlatformTypePtr(models.PlatformTypeNone)}, swag.Bool(true), nil
+		}
+	}
+
+	if *highAvailabilityMode == models.ClusterHighAvailabilityModeNone {
+		if platform != nil && *platform.Type == models.PlatformTypeBaremetal && userManagedNetworking != nil && !*userManagedNetworking {
+			return nil, nil, common.NewApiError(http.StatusBadRequest,
+				errors.Errorf("Can't set %s platform and disable user-managed-networking on single node OpenShift", models.PlatformTypeBaremetal))
+		}
```
Comment (eranco74) [inline internal/bminventory/inventory.go:1762]: No need for this statement (and since it's complicated as it is I think we can remove it), the user will get  badRequest from the next 2 if statements until if he try to set platform=Baremetal or userManagedNetworking=True.

File: internal/bminventory/inventory.go [hunk 13]
```diff
@@ -1668,6 +1676,107 @@ func (b *bareMetalInventory) UpdateClusterNonInteractive(ctx context.Context, pa
...
+			return nil, nil, common.NewApiError(http.StatusBadRequest, errors.Errorf("Can't set %s platform on single node OpenShift", *platform.Type))
+		}
+
+		if userManagedNetworking != nil && !*userManagedNetworking {
+			return nil, nil, common.NewApiError(http.StatusBadRequest, errors.New("Can't disable user-managed-networking on single node OpenShift"))
+		}
+
+		if (userManagedNetworking == nil || *userManagedNetworking) && (platform == nil || *platform.Type == models.PlatformTypeNone) {
+			return &models.Platform{Type: common.PlatformTypePtr(models.PlatformTypeNone)}, swag.Bool(true), nil
+		}
```
Comment (eranco74) [inline internal/bminventory/inventory.go:1774]: Don't even need to check it, just return:
```
return &models.Platform{Type: common.PlatformTypePtr(models.PlatformTypeNone)}, swag.Bool(true), nil
```

File: internal/bminventory/inventory.go [hunk 14]
```diff
@@ -1668,6 +1676,107 @@ func (b *bareMetalInventory) UpdateClusterNonInteractive(ctx context.Context, pa
...
+		// case 4
+		if (userManagedNetworking == nil || *userManagedNetworking) && (platform == nil || *platform.Type == models.PlatformTypeNone) {
+			return nil, nil, nil
+		}
+
+		// case 6
+		if !swag.BoolValue(userManagedNetworking) && (platform == nil || *platform.Type == models.PlatformTypeBaremetal) {
+			if cluster.CPUArchitecture != common.X86CPUArchitecture && !featuresupport.IsFeatureSupported(cluster.OpenshiftVersion, models.FeatureSupportLevelFeaturesItems0FeatureIDARM64ARCHITECTUREWITHCLUSTERMANAGEDNETWORKING) {
+				return nil, nil, common.NewApiError(http.StatusBadRequest, errors.New("disabling User Managed Networking or setting Bare-Metal platform is not allowed for clusters with non-x86_64 CPU architecture"))
+			}
```
Comment (eranco74) [inline internal/bminventory/inventory.go:1729]: Consider moving this to another PR just in case it creates unexpected noise and might get reverted

File: internal/bminventory/inventory.go [hunk 15]
```diff
@@ -1668,6 +1676,107 @@ func (b *bareMetalInventory) UpdateClusterNonInteractive(ctx context.Context, pa
...
+		if (!*userManagedNetworking && *platform.Type == models.PlatformTypeNone) || (*userManagedNetworking && models.PlatformTypeBaremetal == *platform.Type) {
+			return common.NewApiError(http.StatusBadRequest, errors.Errorf("Can't set %s platform with user-managed-networking %s", *platform.Type, userManagedNetworkingStatus))
+		}
+	}
+
+	return nil
+}
+
+func getActualUpdateClusterPlatformParams(platform *models.Platform, userManagedNetworking *bool, cluster *common.Cluster) (*models.Platform, *bool, error) {
+	// case 5
```
Comment (eranco74) [inline internal/bminventory/inventory.go:1695]: Before every return log what we end up with

File: internal/bminventory/inventory.go [hunk 16]
```diff
@@ -1668,6 +1676,107 @@ func (b *bareMetalInventory) UpdateClusterNonInteractive(ctx context.Context, pa
...
+		}
+
+		// case 6
+		if !swag.BoolValue(userManagedNetworking) && (platform == nil || *platform.Type == models.PlatformTypeBaremetal) {
+			if cluster.CPUArchitecture != common.X86CPUArchitecture && !featuresupport.IsFeatureSupported(cluster.OpenshiftVersion, models.FeatureSupportLevelFeaturesItems0FeatureIDARM64ARCHITECTUREWITHCLUSTERMANAGEDNETWORKING) {
+				return nil, nil, common.NewApiError(http.StatusBadRequest, errors.New("disabling User Managed Networking or setting Bare-Metal platform is not allowed for clusters with non-x86_64 CPU architecture"))
+			}
+
+			if *cluster.HighAvailabilityMode == models.ClusterHighAvailabilityModeNone {
+				return nil, nil, common.NewApiError(http.StatusBadRequest, errors.New("disabling User Managed Networking or setting Bare-Metal platform is not allowed in single node mode"))
```
Comment (eranco74) [inline internal/bminventory/inventory.go:1732]:
```suggestion
				return nil, nil, common.NewApiError(http.StatusBadRequest, errors.New("disabling User Managed Networking or setting Bare-Metal platform is not allowed in single node Openshift"))
```

File: internal/bminventory/inventory.go [hunk 17]
```diff
@@ -1668,6 +1676,107 @@ func (b *bareMetalInventory) UpdateClusterNonInteractive(ctx context.Context, pa
...
+			}
+
+			return &models.Platform{Type: common.PlatformTypePtr(models.PlatformTypeBaremetal)}, swag.Bool(false), nil
+		}
+	}
+
+	return platform, userManagedNetworking, nil
+}
+
+func getActualCreateClusterPlatformParams(platform *models.Platform, userManagedNetworking *bool, highAvailabilityMode *string) (*models.Platform, *bool, error) {
```
Comment (eranco74) [inline internal/bminventory/inventory.go:1742]: Before every return log what we end up with

**Review by eranco74 — COMMENTED** [review]
looks good

File: internal/bminventory/inventory.go (hunk 12 above)
Comment (avishayt) [reply internal/bminventory/inventory.go:1762]: The message here is clearer, but not sure it's worth the extra code.  I agree, can remove.

File: internal/bminventory/inventory.go [hunk 18]
```diff
@@ -1668,6 +1676,107 @@ func (b *bareMetalInventory) UpdateClusterNonInteractive(ctx context.Context, pa
...
+		if (platform == nil || *platform.Type == models.PlatformTypeBaremetal) && !swag.BoolValue(userManagedNetworking) {
+			return &models.Platform{Type: common.PlatformTypePtr(models.PlatformTypeBaremetal)}, swag.Bool(false), nil
+		}
+
+		if swag.BoolValue(userManagedNetworking) || (platform != nil && *platform.Type == models.PlatformTypeNone) {
+			return &models.Platform{Type: common.PlatformTypePtr(models.PlatformTypeNone)}, swag.Bool(true), nil
+		}
+	}
+
+	if *highAvailabilityMode == models.ClusterHighAvailabilityModeNone {
```
Comment (avishayt) [inline internal/bminventory/inventory.go:1758]: Why two separate "ifs" instead of an else?

File: internal/bminventory/inventory.go (hunk 15 above)
Comment (avishayt) [reply internal/bminventory/inventory.go:1695]: And please remove all the "// case N" comments

Comment (eliorerz) [reply internal/bminventory/inventory.go:1695]: > And please remove all the "// case N" comments

Sure, that was here just for the first review. I'll remove it.

File: internal/bminventory/inventory.go [hunk 19]
```diff
@@ -1668,6 +1676,107 @@ func (b *bareMetalInventory) UpdateClusterNonInteractive(ctx context.Context, pa
...
+	return platform, userManagedNetworking, nil
+}
+
+func getActualCreateClusterPlatformParams(platform *models.Platform, userManagedNetworking *bool, highAvailabilityMode *string) (*models.Platform, *bool, error) {
+
+	if err := checkPlatformWrongParamsInput(platform, userManagedNetworking); err != nil {
+		return nil, nil, err
+	}
+
+	if *highAvailabilityMode == models.ClusterHighAvailabilityModeFull {
```
[the remark this answers is not in the document]
Comment (eliorerz) [reply internal/bminventory/inventory.go:1748]: It is covered [here](https://github.com/openshift/assisted-service/blob/572b40607fc29728b43a9b418aa2d68d16988760/internal/bminventory/inventory.go#L1749) as part of `!swag.BoolValue(userManagedNetworking)` (it's check if user managed networking is nil or false)

```go
func BoolValue(v *bool) bool {
	if v != nil {
		return *v
	}
	return false
}
```

File: internal/bminventory/inventory.go (hunk 18 above)
Comment (eliorerz) [reply internal/bminventory/inventory.go:1758]: I think it's more clearer to read it this way.
We already have a lot of conditions there and to my opinion it's more clearer to "see" the condition and not need to scroll up to see the if statement.

File: internal/bminventory/inventory.go (hunk 13 above)
Comment (eliorerz) [reply internal/bminventory/inventory.go:1774]: I think that on platforms different than `baremetal` and `none` with `HighAvailabilityMode=None` we still want to return `params.NewClusterParams.Platform` and `params.NewClusterParams.UserManagedNetworking` as they are

Comment (eliorerz) [reply internal/bminventory/inventory.go:1774]: I added a test to check it for vsphere

File: internal/bminventory/inventory.go (hunk 14 above)
Comment (eliorerz) [reply internal/bminventory/inventory.go:1729]: Without it `[It] Fail with non-x86_64 CPU architecture with 4.10` will fail

File: internal/bminventory/inventory.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (avishayt) [reply internal/bminventory/inventory.go:1751]: Right, thanks

File: internal/bminventory/inventory.go (hunk 18 above)
Comment (avishayt) [reply internal/bminventory/inventory.go:1763]: I disagree because this way it's disconnected from the previous "if" even though they are mutually exclusive.  Now I need to scroll up to understand that the code can't actually reach both "ifs".

File: internal/bminventory/inventory.go (hunk 13 above)
Comment (avishayt) [reply internal/bminventory/inventory.go]: SNO is always none platform

File: internal/bminventory/inventory.go (hunk 18 above)
Comment (eliorerz) [reply internal/bminventory/inventory.go:1763]: @avishayt I think that `else if` will solve that issue and still will be more clear, WDYT?

File: internal/bminventory/inventory.go (hunk 13 above)
Comment (eliorerz) [reply internal/bminventory/inventory.go]: Currently, but what about other platforms? For example I don't think that there is any limitation for SNO + Nutanix

Comment (avishayt) [reply internal/bminventory/inventory.go]: @eranco74 is that possible?

File: internal/bminventory/inventory.go (hunk 18 above)
Comment (avishayt) [reply internal/bminventory/inventory.go:1763]: OK

File: internal/bminventory/inventory.go (hunk 13 above)
Comment (eranco74) [reply internal/bminventory/inventory.go]: You can install SNO on Nutanix but the platform will be None!

Comment (eranco74) [reply internal/bminventory/inventory.go]: In summary, in case the cluster HighAvailabilityMode is None, any changes to the UMN or platform that result in a platform other than None should return a bad request.

Comment (eliorerz) [reply internal/bminventory/inventory.go]: Thanks. I'll change it

Comment (eliorerz) [reply internal/bminventory/inventory.go]: @avishayt @eranco74 I moved the if block that checks if the cluster `HighAvailabilityMode==Full` and `Platform!=None` outside the `platform == nil || *platform.Type == models.PlatformTypeBaremetal` block.
Now the service will always return `BedRequest` when trying to set platform != none on SNO

File: internal/bminventory/inventory.go [hunk 20]
```diff
@@ -1668,6 +1676,113 @@ func (b *bareMetalInventory) UpdateClusterNonInteractive(ctx context.Context, pa
...
+		if (platform == nil || *platform.Type == models.PlatformTypeBaremetal) && !swag.BoolValue(userManagedNetworking) {
+			log.Info("Cluster high-availability-mode is set to Full, setting platform type to baremetal and disable user-managed-networking")
+			return &models.Platform{Type: common.PlatformTypePtr(models.PlatformTypeBaremetal)}, swag.Bool(false), nil
+		}
+
+		if swag.BoolValue(userManagedNetworking) || (platform != nil && *platform.Type == models.PlatformTypeNone) {
+			log.Info("Cluster high-availability-mode is set to Full, setting platform type to none and enable user-managed-networking")
+			return &models.Platform{Type: common.PlatformTypePtr(models.PlatformTypeNone)}, swag.Bool(true), nil
+		}
+	} else if *highAvailabilityMode == models.ClusterHighAvailabilityModeNone {
```
Comment (avishayt) [inline internal/bminventory/inventory.go:1765]:
```suggestion
	} else {  // if *highAvailabilityMode == models.ClusterHighAvailabilityModeNone
```

File: internal/bminventory/inventory.go [hunk 21]
```diff
@@ -1668,6 +1676,113 @@ func (b *bareMetalInventory) UpdateClusterNonInteractive(ctx context.Context, pa
...
+			}
+		}
+	}
+
+	log.Infof("No platform manipulation is needed. Setting platform=%v and userManagedNetworking=%v", platform, userManagedNetworking)
+	return platform, userManagedNetworking, nil
+}
+
+func getActualCreateClusterPlatformParams(log logrus.FieldLogger, platform *models.Platform, userManagedNetworking *bool, highAvailabilityMode *string) (*models.Platform, *bool, error) {
+
```
Comment (avishayt) [inline internal/bminventory/inventory.go:1749]:
```suggestion
```

File: internal/bminventory/inventory.go [hunk 22]
```diff
@@ -1668,6 +1676,113 @@ func (b *bareMetalInventory) UpdateClusterNonInteractive(ctx context.Context, pa
...
+			err := errors.New("Can't disable user-managed-networking on single node OpenShift")
+			log.Error(err)
+			return nil, nil, common.NewApiError(http.StatusBadRequest, err)
+		}
+
+		log.Info("Cluster high-availability-mode is set to None, setting platform type to none and enable user-managed-networking")
+		return &models.Platform{Type: common.PlatformTypePtr(models.PlatformTypeNone)}, swag.Bool(true), nil
+	}
+
+	log.Infof("No platform manipulation is needed. Setting platform=%v and userManagedNetworking=%v", platform, userManagedNetworking)
```
Comment (avishayt) [inline internal/bminventory/inventory.go:1782]: This isn't really accurate because there are several cases above where the function returned the same thing that was passed.  I recommend logging the input to the function and the output.

File: internal/bminventory/inventory.go [hunk 23]
```diff
@@ -1668,6 +1676,113 @@ func (b *bareMetalInventory) UpdateClusterNonInteractive(ctx context.Context, pa
...
+					return nil, nil, common.NewApiError(http.StatusBadRequest, err)
+				}
+
+				log.Infof("Updating %s platform to baremetal platform", *cluster.Platform.Type)
+				return &models.Platform{Type: common.PlatformTypePtr(models.PlatformTypeBaremetal)}, swag.Bool(false), nil
+			}
+		}
+	}
+
+	log.Infof("No platform manipulation is needed. Setting platform=%v and userManagedNetworking=%v", platform, userManagedNetworking)
```
Comment (avishayt) [inline internal/bminventory/inventory.go:1744]: Same comment above.  Just log the inputs and outputs for all cases.

File: internal/bminventory/inventory.go [hunk 24]
```diff
@@ -1668,6 +1676,113 @@ func (b *bareMetalInventory) UpdateClusterNonInteractive(ctx context.Context, pa
...
+			return nil, nil, nil
+		}
+
+		if (platform != nil && *platform.Type == models.PlatformTypeNone) || (swag.BoolValue(userManagedNetworking) && platform == nil) {
+			log.Infof("Updating %s platform to none platform", *cluster.Platform.Type)
+			return &models.Platform{Type: common.PlatformTypePtr(models.PlatformTypeNone)}, swag.Bool(true), nil
+		}
+	}
+
+	if *cluster.Platform.Type == models.PlatformTypeNone {
```
Comment (avishayt) [inline internal/bminventory/inventory.go:1716]:
```suggestion
	else if *cluster.Platform.Type == models.PlatformTypeNone {
```

Comment (avishayt) [inline internal/bminventory/inventory.go:1716]:
```suggestion
	} else if *cluster.Platform.Type == models.PlatformTypeNone {
```

**Review by avishayt — APPROVED** [review]
